eras/alonzo/impl/src/Cardano/Ledger/Alonzo/UTxO.hs

```diff
@@ -208,38 +220,81 @@ getInputDataHashesTxBody (UTxO mp) txBody (ScriptsProvided scriptsProvided) =
 -- that every script credential refers to some actual script.  This is tested in the test
 -- function 'validateMissingScripts' in the Utxow rule.
 getAlonzoScriptsNeeded ::
-  forall era.
-  (MaryEraTxBody era, PlutusPurpose AsItem era ~ AlonzoPlutusPurpose AsItem era) =>
+  (MaryEraTxBody era, AlonzoEraScript era) =>
   UTxO era ->
   TxBody era ->
   AlonzoScriptsNeeded era
-getAlonzoScriptsNeeded (UTxO utxo) txBody =
-  AlonzoScriptsNeeded (spend ++ reward ++ cert ++ minted)
+getAlonzoScriptsNeeded utxo txBody =
+  getSpendingScriptsNeeded utxo txBody
+    <> getRewardingScriptsNeeded txBody
+    <> certifyingScriptsNeeded
+    <> getMintingScriptsNeeded txBody
   where
-    collect ::
-      TxIn (EraCrypto era) ->
-      Maybe (AlonzoPlutusPurpose AsItem era, ScriptHash (EraCrypto era))
-    collect !txIn = do
-      addr <- view addrTxOutL <$> Map.lookup txIn utxo
-      hash <- getScriptHash addr
-      return (AlonzoSpending (AsItem txIn), hash)
+    certifyingScriptsNeeded =
+      AlonzoScriptsNeeded $
+        case foldl' addUniqueTxCertPurpose (Map.empty, 0, []) (txBody ^. certsTxBodyL) of
+          (_, _, certPurposes) -> reverse certPurposes
+      where
+        -- We need to do this funny index manipulation here because we've allowed
+        -- duplicate certificates all the way until Conway. This prevented second
+        -- occurance of a duplicate certificate in the sequence to be used. In order to
+        -- preserve this behavior we need to use the index of the first occurance of a
+        -- duplicate certificate.
+        addUniqueTxCertPurpose (!certScriptHashes, !ix, !certPurposes) txCert =
```

Comment:
I struggle to get what's happening here. 
I understand that we're trying to keep the behaviour caused by the possibility of duplicate certificates - but what should actually happen if  that's the case? 
"prevented second occurrence of a duplicate certificate to be used" , so this means that it would always return the first one, as per `indexOf` , no? 
So, in  this implementation,  i can tell we're checking for hashes that we have already seen (identifying duplicates), but I don't understand the "ix + 1" part and what actually is the indexing result  in case of duplicates.  The code is simple enough, so it's just me not getting it, sorry!

Reply:
The `ix + 1` part is to count the actual index of each element. It is only when we see a duplicate we use the index of the first occurrence of the element, but that should not affect indices of other elements. For example:

```haskell
-- these are our certificates
cert = [c0, c1, c2, c3, c4, c5]
```

Let's say `c3` is locked by a native script or a key witness, so it does not participate in the script purpose

Also, let's say `c1 == c4`. Here is what we should get for the plutus purpose:
```haskell
-- these are our certificates
plutusPurpose = [(0, c0), (1, c1), (2, c2), (1, c4), (5, c5)]
```

You see, the count must continue no matter what, thus the counter `ix + 1`, but whenever we find a duplicate we use the stored `ix'`.

Does that help?

Reply:
Yes!  Thank you for taking the time to explain!